### PR TITLE
Table aggregations

### LINKED ISSUE #778: KSQL should support aggregations on tables
Edit: We have a separate (child) ticket that tracks supporting aggregations on non-windowed tables, see https://github.com/confluentinc/ksql/pull/1136.

----

Say you want to count the page views per user across multiple time windows. You may have a `pageviews` stream which has fields like `userid`, `pageid`, etc. for every page view in your system.

If you want a view of activity by user every 30 minutes and also every day, you may do something like this:

```sql
CREATE TABLE user_activity_30_mins AS
  SELECT userid, COUNT(*) AS num_pageviews
  FROM pageviews
  WINDOW TUMBLING (SIZE 30 MINUTES)
  GROUP BY userid;

CREATE TABLE user_activity_24_hr AS
  SELECT userid, sum(num_pageviews)
  FROM user_activity_30_mins
  WINDOW TUMBLING (SIZE 24 HOURS)
  GROUP BY userid;
```

The above won't work, because internally the first query will emit multiple counts per window, and the second query would sum them all, thus over counting. If this was a single streams app and hence a single topology, Kafka Streams will emit `undo` messages whenever a new value is emitted in a given window, hence allowing the downstream `sum` to still be correct.

However, in KSQL both queries are different streams applications, and hence the undo records are not emitted, resulting in over counting.

The above behavior is unintuitive. It means that any kind of 'roll up' application will have wrong results. We need to explore ways to to ensure that only one aggregate is emitted per window, or that all queries in an 'group' behave like a single streams application and enjoy the benefits of 'undo' records, etc.

### PR DESCRIPTION
**Pull request by rodesai**
This patch implements support for aggregating tables without windows. SchemaKTable
now implements groupBy, and returns a SchemaKGroupedTable which can be aggregated
if the requested aggregate functions support subtraction, and if no windowing
expression is specified.

Fixes #778

### CODE REVIEW

File: ksql-engine/src/main/java/io/confluent/ksql/structured/SchemaKStream.java [hunk 1]
```diff
@@ -286,6 +286,40 @@ private boolean rekeyRequired(List<Expression> groupByExpressions) {
...
+  protected static Pair<String, List<Integer>> keyIndexesForGroupBy(
+      final Schema schema, final List<Expression> groupByExpressions) {
+    final StringBuilder aggregateKeyName = new StringBuilder();
+    final List<Integer> newKeyIndexes = new ArrayList<>();
+
+    // Collect the column indexes, and build the new key as <column1>+<column2>+...
+    boolean addSeparator = false;
+    for (Expression groupByExpr : groupByExpressions) {
+      if (addSeparator) {
+        aggregateKeyName.append("|+|");
```
Comment (hjafarpour) [inline ksql-engine/src/main/java/io/confluent/ksql/structured/SchemaKStream.java]: Now that we are using `|+|` in more than one place maybe yo can make it a final static variable.

File: ksql-engine/src/main/java/io/confluent/ksql/function/KsqlAggregateFunction.java [hunk 2]
```diff
@@ -62,6 +62,14 @@ public boolean hasSameArgTypes(List<Schema> argTypeList) {
 
   public abstract A aggregate(V currentVal, A currentAggVal);
 
+  public A subtract(V currentVal, A currentAggVal) {
```
Comment (hjafarpour) [inline ksql-engine/src/main/java/io/confluent/ksql/function/KsqlAggregateFunction.java]: Maybe change the name to `undo`?

File: ksql-engine/src/main/java/io/confluent/ksql/function/udaf/count/CountKudaf.java [hunk 3]
```diff
@@ -44,6 +44,16 @@ public Long aggregate(Object currentVal, Long currentAggVal) {
     return (aggKey, aggOne, aggTwo) -> aggOne + aggTwo;
   }
 
+  @Override
+  public Long subtract(Object currentVal, Long currentAggVal) {
+    return currentAggVal - 1;
+  }
+
+  @Override
+  public boolean implementsSubtract() {
```
Comment (dguy) [inline ksql-engine/src/main/java/io/confluent/ksql/function/udaf/count/CountKudaf.java]: should we do this with an interface instead?

File: ksql-engine/src/main/java/io/confluent/ksql/function/KsqlAggregateFunction.java (hunk 2 above)
Comment (dguy) [inline ksql-engine/src/main/java/io/confluent/ksql/function/KsqlAggregateFunction.java]: maybe have another interface for this?

File: ksql-engine/src/main/java/io/confluent/ksql/function/udaf/KudafSubtractor.java [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+import io.confluent.ksql.function.KsqlAggregateFunction;
+import org.apache.kafka.streams.kstream.Aggregator;
+
+import java.util.Map;
+
+public class KudafSubtractor implements Aggregator<String, GenericRow, GenericRow> {
+  private Map<Integer, KsqlAggregateFunction> aggValToAggFunctionMap;
+  private Map<Integer, Integer> aggValToValColumnMap;
+
+  public KudafSubtractor(Map<Integer, KsqlAggregateFunction> aggValToAggFunctionMap, Map<Integer,
```
Comment (dguy) [inline ksql-engine/src/main/java/io/confluent/ksql/function/udaf/KudafSubtractor.java]: nit: arguments on new line

File: ksql-engine/src/main/java/io/confluent/ksql/function/udaf/KudafSubtractor.java [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+      Integer> aggValToValColumnMap) {
+    this.aggValToAggFunctionMap = aggValToAggFunctionMap;
+    this.aggValToValColumnMap = aggValToValColumnMap;
+  }
+
+  @SuppressWarnings("unchecked")
+  @Override
+  public GenericRow apply(String s, GenericRow rowValue, GenericRow aggRowValue) {
```
Comment (dguy) [inline ksql-engine/src/main/java/io/confluent/ksql/function/udaf/KudafSubtractor.java]: unit test for this?

File: ksql-engine/src/main/java/io/confluent/ksql/structured/SchemaKGroupedStream.java [hunk 6]
```diff
@@ -38,12 +41,12 @@
 
 public class SchemaKGroupedStream {
 
-  private final Schema schema;
-  private final KGroupedStream kgroupedStream;
-  private final Field keyField;
-  private final List<SchemaKStream> sourceSchemaKStreams;
-  private final FunctionRegistry functionRegistry;
-  private final SchemaRegistryClient schemaRegistryClient;
+  protected final Schema schema;
```
Comment (dguy) [inline ksql-engine/src/main/java/io/confluent/ksql/structured/SchemaKGroupedStream.java]: Can we make these package-private?

File: ksql-engine/src/main/java/io/confluent/ksql/structured/SchemaKGroupedTable.java [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+import org.apache.kafka.streams.kstream.Initializer;
+import org.apache.kafka.streams.kstream.KGroupedTable;
+import org.apache.kafka.streams.kstream.KTable;
+import org.apache.kafka.streams.kstream.Materialized;
+
+import java.util.List;
+import java.util.Map;
+
+public class SchemaKGroupedTable extends SchemaKGroupedStream {
+  KGroupedTable kgroupedTable;
```
Comment (dguy) [inline ksql-engine/src/main/java/io/confluent/ksql/structured/SchemaKGroupedTable.java]: `private final`

File: ksql-engine/src/main/java/io/confluent/ksql/structured/SchemaKGroupedTable.java [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+      Map<Integer, Integer> aggValToValColumnMap,
+      final WindowExpression windowExpression,
+      final Serde<GenericRow> topicValueSerDe) {
+    if (windowExpression != null) {
+      throw new KsqlException("Windowing not supported for table aggregations.");
+    }
+    if (aggValToFunctionMap.values()
+            .stream()
+            .map(e -> !e.implementsSubtract())
+            .reduce(false, (r, e) ->  r || e)) {
```
Comment (dguy) [inline ksql-engine/src/main/java/io/confluent/ksql/structured/SchemaKGroupedTable.java]: can we use real variable names?

File: ksql-engine/src/main/java/io/confluent/ksql/structured/SchemaKStream.java [hunk 9]
```diff
@@ -286,6 +286,40 @@ private boolean rekeyRequired(List<Expression> groupByExpressions) {
...
+        addSeparator = true;
+      }
+      aggregateKeyName.append(groupByExpr.toString());
+      newKeyIndexes.add(
+          SchemaUtil.getIndexInSchema(groupByExpr.toString(), schema));
+    }
+    return new Pair<>(aggregateKeyName.toString(), newKeyIndexes);
+  }
+
+  protected static String buildGroupByKey(List<Integer> newKeyIndexes, GenericRow value) {
```
Comment (dguy) [inline ksql-engine/src/main/java/io/confluent/ksql/structured/SchemaKStream.java]: package-private ?

File: ksql-engine/src/main/java/io/confluent/ksql/structured/SchemaKTable.java [hunk 10]
```diff
@@ -174,4 +176,27 @@ public boolean isWindowed() {
     return isWindowed;
   }
 
+  @Override
+  public SchemaKGroupedStream groupBy(
```
Comment (dguy) [inline ksql-engine/src/main/java/io/confluent/ksql/structured/SchemaKTable.java]: nit: new line for each param

File: ksql-engine/src/main/java/io/confluent/ksql/function/KsqlUndoableAggregationFunction.java [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+public abstract class KsqlUndoableAggregationFunction<V, A> extends KsqlAggregateFunction<V, A> {
+  public KsqlUndoableAggregationFunction(
+      final int argIndexInValue,
+      final Supplier<A> initialValueSupplier,
+      final Schema returnType,
+      final List<Schema> argument) {
+    super(argIndexInValue, initialValueSupplier, returnType, argument);
+  }
+
+  public abstract A undo(V currentVal, A currentAggVal);
```
Comment (big-andy-coates) [inline ksql-engine/src/main/java/io/confluent/ksql/function/KsqlUndoableAggregationFunction.java]: shouldn't `currentVal` be `oldValue` or `valueToRemove`?  I find `currentVal` confusing, as its not the current value being processed.

File: ksql-engine/src/main/java/io/confluent/ksql/structured/SchemaKStream.java [hunk 12]
```diff
@@ -286,8 +287,43 @@ private boolean rekeyRequired(List<Expression> groupByExpressions) {
...
   }
 
+  static Pair<String, List<Integer>> keyIndexesForGroupBy(
+      final Schema schema, final List<Expression> groupByExpressions) {
+    final StringBuilder aggregateKeyName = new StringBuilder();
+    final List<Integer> newKeyIndexes = new ArrayList<>();
+
+    // Collect the column indexes, and build the new key as <column1>+<column2>+...
+    boolean addSeparator = false;
+    for (Expression groupByExpr : groupByExpressions) {
```
Comment (big-andy-coates) [inline ksql-engine/src/main/java/io/confluent/ksql/structured/SchemaKStream.java]: I think using streams here would be cleaner:

```java
final String aggregateKeyName = groupByExpressions.stream()
        .map(Expression::toString)
        .collect(Collectors.joining(GROUP_BY_COLUMN_SEPARATOR));
    
groupByExpressions.forEach(e -> SchemaUtil.getIndexInSchema(e.toString(), schema));
```

(I realize you've only refactored existing code into a method, so feel free to ignore if you want as you've already improved the code :D

File: ksql-engine/src/main/java/io/confluent/ksql/structured/SchemaKStream.java [hunk 13]
```diff
@@ -286,8 +287,43 @@ private boolean rekeyRequired(List<Expression> groupByExpressions) {
...
+      }
+      aggregateKeyName.append(groupByExpr.toString());
+      newKeyIndexes.add(
+          SchemaUtil.getIndexInSchema(groupByExpr.toString(), schema));
+    }
+    return new Pair<>(aggregateKeyName.toString(), newKeyIndexes);
+  }
+
+  static String buildGroupByKey(List<Integer> newKeyIndexes, GenericRow value) {
+    StringBuilder newKey = new StringBuilder();
```
Comment (big-andy-coates) [inline ksql-engine/src/main/java/io/confluent/ksql/structured/SchemaKStream.java]: Likewise:

```
return newKeyIndexes.stream()
        .map(idx -> String.valueOf(value.getColumns().get(idx)))
        .collect(Collectors.joining(GROUP_BY_COLUMN_SEPARATOR));
```

File: ksql-engine/src/main/java/io/confluent/ksql/structured/SchemaKStream.java [hunk 14]
```diff
@@ -303,40 +339,19 @@ public SchemaKGroupedStream groupBy(
...
-      if (addSeparator) {
-        aggregateKeyName.append("|+|");
-      } else {
-        addSeparator = true;
-      }
-      aggregateKeyName.append(groupByExpr.toString());
-      newKeyIndexes.add(
-          SchemaUtil.getIndexInSchema(groupByExpr.toString(), getSchema()));
-    }
+    final Pair<String, List<Integer>> aggregateKeyNameAndNewKeyIndexes =
```
Comment (big-andy-coates) [inline ksql-engine/src/main/java/io/confluent/ksql/structured/SchemaKStream.java]: Personally, I'm not a fan of `Pair` type classes as it makes the code hard to read.  Concrete specific types lead to much more readable code IMHO.

File: ksql-engine/src/main/java/io/confluent/ksql/structured/SchemaKTable.java [hunk 15]
```diff
@@ -79,9 +81,9 @@ public SchemaKTable into(
     if (isWindowed) {
       final Serde<Windowed<String>> windowedSerde
               = WindowedSerdes.timeWindowedSerdeFrom(String.class);
-      ktable.toStream()
+      ((KTable<Windowed<String>, GenericRow>)ktable).toStream()
```
Comment (big-andy-coates) [inline ksql-engine/src/main/java/io/confluent/ksql/structured/SchemaKTable.java]: You could use two constructors and two private inner class implementations to handle windowed and non-windowed KTables, with the correct generics, e.g. one constructor that took a `KTable<Windowed<String>, GenericRow>` and one that took a `KTable<String, GenericRow>`.

this would make the whole thing more type safe and remove the needs for the casts.

File: ksql-engine/src/test/resources/query-validation-tests/aggregate-table.json [hunk 16]
```diff
@@ -0,0 +1,34 @@
...
+      "statements": [
+        "CREATE TABLE TEST (ID bigint, NAME varchar, REGION string) WITH (kafka_topic='test_topic', value_format='DELIMITED', key='ID');",
+        "CREATE TABLE COUNT_BY_REGION AS SELECT REGION, count(*) FROM TEST GROUP BY REGION;"
+      ],
+      "inputs": [
+        {"topic": "test_topic", "key": 0, "value": "0,alice,east", "timestamp": 0},
+        {"topic": "test_topic", "key": 1, "value": "1,bob,east", "timestamp": 0},
+        {"topic": "test_topic", "key": 2, "value": "2,carol,west", "timestamp": 0},
+        {"topic": "test_topic", "key": 3, "value": "3,dave,west", "timestamp": 0},
+        {"topic": "test_topic", "key": 1, "value": "1,alice,west", "timestamp": 0}
```
Comment (big-andy-coates) [inline ksql-engine/src/test/resources/query-validation-tests/aggregate-table.json]: Consider adding a delete / null-value into the mix just for completeness?

File: ksql-engine/src/main/java/io/confluent/ksql/function/udaf/KudafUndoAggregator.java [hunk 17]
```diff
@@ -0,0 +1,52 @@
...
+import java.util.Map;
+
+public class KudafUndoAggregator implements Aggregator<String, GenericRow, GenericRow> {
+  private Map<Integer, KsqlUndoableAggregationFunction> aggValToAggFunctionMap;
+  private Map<Integer, Integer> aggValToValColumnMap;
+
+  public KudafUndoAggregator(
+      Map<Integer, KsqlUndoableAggregationFunction> aggValToAggFunctionMap,
+      Map<Integer, Integer> aggValToValColumnMap) {
+    this.aggValToAggFunctionMap = aggValToAggFunctionMap;
```
[the remark this answers is not in the document]
Comment (rodesai) [reply ksql-engine/src/main/java/io/confluent/ksql/function/udaf/KudafUndoAggregator.java]: Thanks for calling this out. I totally agree. My background is more C++ where references + const-correctness let you either prevent or check for this kind of thing at compile time. Still adjusting to the java way of defining these kinds of constraints. I'll add null checks and make the maps immutable copies, and go through my other open PRs and do the same.

File: ksql-engine/src/main/java/io/confluent/ksql/function/KsqlUndoableAggregationFunction.java [hunk 18]
```diff
@@ -0,0 +1,34 @@
...
+ **/
+
+package io.confluent.ksql.function;
+
+import org.apache.kafka.connect.data.Schema;
+
+import java.util.List;
+import java.util.function.Supplier;
+
+public abstract class KsqlUndoableAggregationFunction<V, A> extends KsqlAggregateFunction<V, A> {
```
[the remark this answers is not in the document]
Comment (rodesai) [reply ksql-engine/src/main/java/io/confluent/ksql/function/KsqlUndoableAggregationFunction.java]: I did it this way at first. But then I found myself having to cast from the interface to KsqlAggregateFunction to get to getArgIndexInValue. Which made me think that a KsqlUndoableAggregateFunction is really a special type of a KsqlAggregateFunction that also implements an undo method. So I changed it to a child class.

File: ksql-engine/src/main/java/io/confluent/ksql/function/udaf/KudafUndoAggregator.java [hunk 19]
```diff
@@ -0,0 +1,52 @@
...
+    this.aggValToValColumnMap = aggValToValColumnMap;
+  }
+
+  @SuppressWarnings("unchecked")
+  @Override
+  public GenericRow apply(String s, GenericRow rowValue, GenericRow aggRowValue) {
+    aggValToValColumnMap.forEach(
```
[the remark this answers is not in the document]
Comment (rodesai) [reply ksql-engine/src/main/java/io/confluent/ksql/function/udaf/KudafUndoAggregator.java]: The undo aggregation is producing a new row for the key we're removing a value from. I think its easiest to illustrate this with an example:

Consider the following statements
```sql
CREATE TABLE T (A int, B int, C VARCHAR) WITH (KEY='A', ...);
CREATE TABLE TAGG AS SELECT C, SUM(B) GROUP BY C;
```
Lets say the current state of T is:
1 | 5 | B
2 | 7 | B
3 | 3 | A

Then the current state of TAGG is:
A | 3
B | 12

Now let's say the following update comes in for T:
2 | 5 | A

KSQL should then emit the following 2 changelog records for TAGG:
B | 5
A | 8

The update to B is the result of the undo
The update to A is the result of the regular aggregation

File: ksql-engine/src/main/java/io/confluent/ksql/structured/SchemaKGroupedStream.java [hunk 20]
```diff
@@ -38,12 +41,12 @@
 
 public class SchemaKGroupedStream {
 
-  private final Schema schema;
-  private final KGroupedStream kgroupedStream;
-  private final Field keyField;
-  private final List<SchemaKStream> sourceSchemaKStreams;
-  private final FunctionRegistry functionRegistry;
-  private final SchemaRegistryClient schemaRegistryClient;
+  final Schema schema;
```
[the remark this answers is not in the document]
Comment (rodesai) [reply ksql-engine/src/main/java/io/confluent/ksql/structured/SchemaKGroupedStream.java]: These are package-private right now. That's stronger than protected.

File: ksql-engine/src/main/java/io/confluent/ksql/structured/SchemaKGroupedTable.java [hunk 21]
```diff
@@ -0,0 +1,98 @@
...
+import org.apache.kafka.streams.kstream.Initializer;
+import org.apache.kafka.streams.kstream.KGroupedTable;
+import org.apache.kafka.streams.kstream.KTable;
+import org.apache.kafka.streams.kstream.Materialized;
+
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+public class SchemaKGroupedTable extends SchemaKGroupedStream {
```
[the remark this answers is not in the document]
Comment (rodesai) [reply ksql-engine/src/main/java/io/confluent/ksql/structured/SchemaKGroupedTable.java]: I agree we shouldn't use inheritance here. But we're kind of forced to because of how the other Schema.*(Stream|Table) classes are structured. SchemaKTable is a sub-class of SchemaKStream, and the plan nodes only ever deal with SchemaKStreams. So to keep things consistent I mirrored that here. We should probably rethink the relationships between those 4 classes and do a refactor, but I'm not sure that it makes sense to do so in _this_ PR.

File: ksql-engine/src/main/java/io/confluent/ksql/structured/SchemaKGroupedTable.java [hunk 22]
```diff
@@ -0,0 +1,98 @@
...
+  SchemaKGroupedTable(
+      final Schema schema,
+      final KGroupedTable kgroupedTable,
+      final Field keyField,
+      final List<SchemaKStream> sourceSchemaKStreams,
+      final FunctionRegistry functionRegistry,
+      final SchemaRegistryClient schemaRegistryClient
+  ) {
+    super(schema, null, keyField, sourceSchemaKStreams, functionRegistry, schemaRegistryClient);
+    this.kgroupedTable = kgroupedTable;
```
[the remark this answers is not in the document]
Comment (rodesai) [reply ksql-engine/src/main/java/io/confluent/ksql/structured/SchemaKGroupedTable.java]: I'll use Objects.requireNonNull. I don't think it makes sense to check it and mask the NPE with some other exception though.

File: ksql-engine/src/main/java/io/confluent/ksql/structured/SchemaKStream.java (hunk 12 above)
Comment (rodesai) [reply ksql-engine/src/main/java/io/confluent/ksql/structured/SchemaKStream.java]: Makes sense to me. I'll write it with a stream.

File: ksql-engine/src/main/java/io/confluent/ksql/function/udaf/sum/IntegerSumKudaf.java [hunk 23]
```diff
@@ -42,6 +43,11 @@ public Integer aggregate(Integer currentVal, Integer currentAggVal) {
     return currentVal + currentAggVal;
   }
 
+  @Override
+  public Integer undo(Integer currentVal, Integer currentAggVal) {
```
[the remark this answers is not in the document]
Comment (rodesai) [reply ksql-engine/src/main/java/io/confluent/ksql/function/udaf/sum/IntegerSumKudaf.java]: Actually this won't work since you can't use operators on generic types or Number

File: ksql-engine/src/main/java/io/confluent/ksql/structured/SchemaKStream.java (hunk 14 above)
Comment (rodesai) [reply ksql-engine/src/main/java/io/confluent/ksql/structured/SchemaKStream.java]: I'll just split this into 2 functions

File: ksql-engine/src/main/java/io/confluent/ksql/structured/SchemaKTable.java (hunk 15 above)
Comment (rodesai) [reply ksql-engine/src/main/java/io/confluent/ksql/structured/SchemaKTable.java]: That won't work because both methods would have the same erasure. We could use static factory functions, but even then they wouldn't actually check anything since we're passing around raw KTables/KStreams/KGroupedStreams/KGroupedTables in most places. In the interest of keeping this PR efficient how about I file an issue for actually using type parameters with these generics and enforcing type safety?

File: ksql-engine/src/test/resources/query-validation-tests/aggregate-table.json (hunk 16 above)
Comment (rodesai) [reply ksql-engine/src/test/resources/query-validation-tests/aggregate-table.json]: Thanks for the suggestion here. Looks like we have a general bug with tables where we drop tombstones on the floor. I'll file an issue and fix that separately since we should probably get that in asap.

Comment (rodesai) [reply ksql-engine/src/test/resources/query-validation-tests/aggregate-table.json]: https://github.com/confluentinc/ksql/issues/1169

File: ksql-engine/src/main/java/io/confluent/ksql/function/udaf/KudafUndoAggregator.java [hunk 24]
```diff
@@ -0,0 +1,58 @@
...
+        (aggRowIndex, rowIndex) ->
+            aggRowValue.getColumns().set(aggRowIndex, rowValue.getColumns().get(rowIndex)));
+
+    aggValToAggFunctionMap.forEach(
+        (aggRowIndex, function) ->
+            aggRowValue.getColumns().set(
+                aggRowIndex,
+                function.undo(
+                    rowValue.getColumns().get(function.getArgIndexInValue()),
+                    aggRowValue.getColumns().get(aggRowIndex))));
```
Comment (apurvam) [inline ksql-engine/src/main/java/io/confluent/ksql/function/udaf/KudafUndoAggregator.java]: Not sure I follow what's going on here. The `aggRowValue[aggRowIndex]` is set to `rowValue[rowIndex]` on line 47. 

on this line, we are simply performing an aggregate on the value, which makes me wonder what purpose the operation on line 47 is. Doesn't the operation on line 47 clobber the existing aggregation value? It _seems_ to be setting the aggregation value to the incoming value. Why is it correct?

(I am confident it's correct, I am asking to clarify my understanding.)

I know this pattern is copied over from the existing aggregation function, so it may make sense to have a comment in the original pattern to explain what is going on here.

File: ksql-engine/src/main/java/io/confluent/ksql/function/udaf/sum/IntegerSumKudaf.java [hunk 25]
```diff
@@ -35,11 +36,16 @@
...
+    if (currentValue == null) {
+      return aggregateValue;
     }
-    return currentVal + currentAggVal;
+    return currentValue + aggregateValue;
+  }
+
+  @Override
+  public Integer undo(Integer valueToUndo, Integer aggregateValue) {
+    return this.aggregate(-1 * valueToUndo, aggregateValue);
```
Comment (apurvam) [inline ksql-engine/src/main/java/io/confluent/ksql/function/udaf/sum/IntegerSumKudaf.java]: The `DoubleSumKudaf` just does a plain subtraction. Why is this different? Perhaps add a comment to the code if there is a good reason for the discrepancy?

File: ksql-engine/src/main/java/io/confluent/ksql/physical/AddTimestampColumn.java [hunk 26]
```diff
@@ -36,6 +36,7 @@ public void init(ProcessorContext processorContext) {
       public GenericRow transform(GenericRow row) {
         if (row != null) {
           row.getColumns().add(0, processorContext.timestamp());
+          return row;
```
Comment (apurvam) [inline ksql-engine/src/main/java/io/confluent/ksql/physical/AddTimestampColumn.java]: Is this intentional? Seems pretty pointless to add this.

File: ksql-engine/src/main/java/io/confluent/ksql/planner/plan/AggregateNode.java [hunk 27]
```diff
@@ -143,7 +142,7 @@ public WindowExpression getWindowExpression() {
   private List<Pair<String, Expression>> getFinalSelectExpressions() {
     List<Pair<String, Expression>> finalSelectExpressionList = new ArrayList<>();
     if (finalSelectExpressions.size() != schema.fields().size()) {
-      throw new KsqlException(
+      throw new RuntimeException(
```
Comment (apurvam) [inline ksql-engine/src/main/java/io/confluent/ksql/planner/plan/AggregateNode.java]: Why change the type of exception here?

File: ksql-engine/src/main/java/io/confluent/ksql/planner/plan/AggregateNode.java [hunk 28]
```diff
@@ -368,39 +357,6 @@ private Schema buildAggregateSchema(
     return schemaBuilder.build();
   }
 
-  private SchemaKStream maybeSetTheKeyField(SchemaKStream schemaKStream,
```
Comment (apurvam) [inline ksql-engine/src/main/java/io/confluent/ksql/planner/plan/AggregateNode.java]: Why don't we need this any longer?

File: ksql-engine/src/main/java/io/confluent/ksql/structured/SchemaKStream.java [hunk 29]
```diff
@@ -127,65 +129,103 @@ public SchemaKStream filter(final Expression filterExpression) {
...
-        schemaAndMapper.left,
-        kstream.mapValues(schemaAndMapper.right),
-        keyField,
-        Collections.singletonList(this),
-        Type.PROJECT,
-        functionRegistry,
-        schemaRegistryClient
-    );
-  }
+    private Field findKeyField(
```
Comment (apurvam) [inline ksql-engine/src/main/java/io/confluent/ksql/structured/SchemaKStream.java]: Why introduce this method? Is necessary for `SchemaKTable` and nice to have for `SchemaKStream`?

File: ksql-engine/src/main/java/io/confluent/ksql/physical/AddTimestampColumn.java (hunk 26 above)
Comment (rodesai) [reply ksql-engine/src/main/java/io/confluent/ksql/physical/AddTimestampColumn.java]: Good catch - I moved some of the code around so I could set a breakpoint for the null case. I'll revert.

File: ksql-engine/src/main/java/io/confluent/ksql/planner/plan/AggregateNode.java (hunk 27 above)
Comment (rodesai) [reply ksql-engine/src/main/java/io/confluent/ksql/planner/plan/AggregateNode.java]: IMO at some point we should have an exception hierarchy that at minimum categorizes our exceptions as being problems with inputs, external system errors, or internal errors (e.g. broken assumptions). I think for now we should try and stick to using KsqlException to indicate an input error (e.g. something was wrong with a statement) since that's overwhelmingly the most common reason KsqlException is raised. (Currently the only place where we try and make that distinction is the rest layer, which returns 400 when it sees a KsqlException). This condition is more like an internal error, so I changed it to raise something other than KsqlException.

File: ksql-engine/src/main/java/io/confluent/ksql/planner/plan/AggregateNode.java (hunk 28 above)
Comment (rodesai) [reply ksql-engine/src/main/java/io/confluent/ksql/planner/plan/AggregateNode.java]: See below comment

File: ksql-engine/src/main/java/io/confluent/ksql/structured/SchemaKStream.java (hunk 29 above)
Comment (rodesai) [reply ksql-engine/src/main/java/io/confluent/ksql/structured/SchemaKStream.java]: This change encapsulates management of the key field away from the client code. So if the select call renames the key field then the SchemaKStream/SchemaKTable figures out the new key field and sets it correctly in the returned SchemaKStream/SchemaKTable.

This way AggregateNode doesn't need the maybeSetTheKeyField stuff that creates a new SchemaKStream with the key field that uses internal names. I think in general we should try and avoid constructing these SchemaKStream and SchemaKTable objects from plan nodes (maybe even going as far as making the constructors package-private and providing a static factory function for just StructuredDataSourceNode to use). This avoids the problem of accidentally converting from a SchemaKTable to a SchemaKStream.

File: ksql-engine/src/main/java/io/confluent/ksql/function/udaf/KudafUndoAggregator.java (hunk 24 above)
Comment (rodesai) [reply ksql-engine/src/main/java/io/confluent/ksql/function/udaf/KudafUndoAggregator.java]: The first statement is copying all the grouping expressions and arguments into the row. The second statement does the aggregation and writes that into the row.

File: ksql-engine/src/test/java/io/confluent/ksql/structured/SchemaKTableTest.java [hunk 30]
```diff
@@ -162,4 +186,62 @@ public void testFilter() throws Exception {
                       initialSchemaKTable);
   }
 
+  @Test
+  public void testGroupBy() {
+    String selectQuery = "SELECT col0, col1, col2 FROM test1;";
```
Comment (apurvam) [inline ksql-engine/src/test/java/io/confluent/ksql/structured/SchemaKTableTest.java]: Isn't test1 a stream? Should this be test2? Or are you testing groupBy on streams? Or does it not matter what you are selecting from.. all you are interested in is the schema of the source node?

File: ksql-engine/src/test/java/io/confluent/ksql/structured/SchemaKTableTest.java [hunk 31]
```diff
@@ -162,4 +186,62 @@ public void testFilter() throws Exception {
...
+    SchemaKGroupedStream groupedSchemaKTable = initialSchemaKTable.groupBy(
+        Serdes.String(), rowSerde, groupByExpressions);
+
+    assertThat(groupedSchemaKTable, instanceOf(SchemaKGroupedTable.class));
+    assertThat(groupedSchemaKTable.getKeyField().name(), equalTo("TEST1.COL2|+|TEST1.COL1"));
+  }
+
+  @SuppressWarnings("unchecked")
+  @Test
+  public void shouldGroupKeysCorrectly() {
```
Comment (apurvam) [inline ksql-engine/src/test/java/io/confluent/ksql/structured/SchemaKTableTest.java]: I think it would really help in tests like this to list out the givens and expectation in words. Its really hard to follow what you are trying to test from the code, with all the mocks and intermediate steps involved.

File: ksql-engine/src/main/java/io/confluent/ksql/planner/plan/AggregateNode.java (hunk 27 above)
Comment (rodesai) [reply ksql-engine/src/main/java/io/confluent/ksql/planner/plan/AggregateNode.java]: That's correct

File: ksql-engine/src/test/java/io/confluent/ksql/structured/SchemaKTableTest.java (hunk 30 above)
Comment (rodesai) [reply ksql-engine/src/test/java/io/confluent/ksql/structured/SchemaKTableTest.java]: Good catch. Copypasta from the other tests. The tests are still valid since the schema happens to be compatible. But we should use the right source here. Fixing it.

File: ksql-engine/src/main/java/io/confluent/ksql/function/KsqlUndoableAggregationFunction.java (hunk 18 above)
[the remark this answers is not in the document]
Comment (dguy) [reply ksql-engine/src/main/java/io/confluent/ksql/function/KsqlUndoableAggregationFunction.java]: Seems like an interface to me, too. I think we also need an interface for `KsqlAggregateFunction` then perhaps we wouldn't need this inheritance? i.e, if we had 2 interfaces and the interface for this extended the default aggregate interface?

File: ksql-engine/src/main/java/io/confluent/ksql/function/KsqlUndoableAggregationFunction.java [hunk 32]
```diff
@@ -0,0 +1,34 @@
...
+  public KsqlUndoableAggregationFunction(
```
Comment (dguy) [inline ksql-engine/src/main/java/io/confluent/ksql/function/KsqlUndoableAggregationFunction.java]: Perhaps `TableAggregationFunction`?

or just `UndoableAggregationFunction`, i.e., we don't need to name everything `Ksql...` as it makes it a pain when you are looking for classes if everything starts with the same thing. We know it is `Ksql`...

**Comment by miguno** [comment]
Thanks @rodesai !
